### DISCUSSION TITLE
VSCode-Cpptools/ipch/folders. should i delete them? or are they necessary everytime i use it?

### DISCUSSION THREAD

**Discussion by Keanu-Type** [microsoft/vscode-cpptools#10637] [Q&A] [answered]
using TreeSize Free i manage to delete and locate unnecessary files(previously undeleted game files, cache files from uninstalled apps, etc.)  taking too much space in my disks. Then finally i came across to this:
In my **C:/Users/PCNAME/AppData/Local/Microsoft/vscode-cpptools/ipch/** there are a LOT of folders here. it is currently at ~5GB. However i saved up my .cpp, .c, .h files in **D:/Documents/Cplusplus** directory. 

Additional context, whenever i made and try to compile a file, i always prefer doing it via Bash commandline such as "g++ filename.cpp -o finalfilename". not sure if i should mention this but worth adding if needed a better context on how i use C++ in my VSCode.

So i have a questions:
1.  Are this "Cache" files for the .cpp, .c, .h files i made? judging by the Last Modified, this are files and folders when i was doing C++ coding like reviewing back my knowledge/watching C++ tutorials back. so i was assuming it is...
2. Are this file inside the **Ipch/** safe to delete? since im currently not doing C++ can I safely delete them to save up some space?
 2a. if i make a new C++ file/project, will they generate a new one for that file?
2b. if i reopen and run an old C++ file/s, will they regenerate a new copy or will it cause an error since (assuming) i deleted those "Cache" folders/files from **Ipch/** directory?

I hope I can ask this questions since i currently wanted to know and clarify what are this folders for. So in the future i know what can i remove and retain in, but ultimately is to learn their purpose. 
Best Regards, Keanu-Type

![image](https://user-images.githubusercontent.com/72327819/223126717-6a474877-2fb8-4b8a-8b52-468270fb9e2d.png)

**Comment by sean-mcmanus** [comment]
The ipch cache folder can be safely deleted -- it's just used to speed up IntelliSense for any C/C++ files that are opened. It's better to delete the folder after closing VS Code though, in case one of our extension processes is using a cache file.

**Reply by Keanu-Type** [reply]
Thank you for clarifying that, good sir!
Best regards, Keanu-Type

**Comment by nouran-19** [comment]
what do you mean by "saved up my .cpp, .c, .h files" 
I mean did you just move the code files to another directory?

**Comment by TryerGit** [comment]
> The ipch cache folder can be safely deleted -- it's just used to speed up IntelliSense for any C/C++ files that are opened. It's better to delete the folder after closing VS Code though, in case one of our extension processes is using a cache file.

As developers of this extension, could you please consider providing an exhaustive and complete list of folders and files that can be safely deleted without breaking any functionality? Then, it is up to the user to decide to delete the ipch files as and when they are comfortable doing so.

It is quite strange that users have to ask why there are so many files and folders created all over the place which are multiple GBs in size. These files add up quickly and hard disk space is at a premium.

I have posted this question with Visual Studio IDE as well (which also has the sneaky tendency to create arbitrary files multiple GBs in size all over the place) but have had trouble figuring out the exact files that can be safely deleted.

**Reply by bobbrow** [reply, marked as answer]
All of the files we create can be safely deleted.  They are caches that exist to make IntelliSense features faster.  If we look for them and they are gone we will simply re-create them.

Any given workspace can override where our databases and IntelliSense caches (ipch) are written to, but in the absence of such an override, they will go here:
**Windows** - `%localappdata%/Microsoft/vscode-cpptools`
**macOS** - `~/Library/Caches/vscode-cpptools`
**Linux** - `$XDG_CACHE_HOME/vscode-cpptools` or `~/.cache/vscode-cpptools` if the env variable does not exist

Reference: https://github.com/microsoft/vscode-cpptools/blob/ae8c4aca58599dfa6d6d39849cadbfa7c5af52a1/Extension/src/common.ts#L1312-L1334

**Reply by Self4215** [reply]
> I have posted this question with Visual Studio IDE as well (which also has the sneaky tendency to create arbitrary files multiple GBs in size all over the place) but have had trouble figuring out the exact files that can be safely deleted.

For Visual Studio, perhaps it's due to the same ipch issue. Navigate to `Tools - Options - Text Editor - C/C++ - Advanced - Intellisense - Disable Automatic Precompiled Header`, and set it to True. Then, you can delete this folder for each solution: `<solution folder>/.vs/<solution name>/v17/ipch` (you need to enable "Show hidden files, folders, and drives" in your file explorer).

**Comment by mrmbernardi** [comment]
Is there a setting we can configure somewhere to reduce the size of the cache?

**Reply by bobbrow** [reply]
Yes. It's this one:
![image](https://github.com/microsoft/vscode-cpptools/assets/12818240/fd3b4e7a-799d-413c-bd9f-7b12dee2acba)

**Reply by joebed** [reply]
How does this work? My Cache size as 5120, but my ipch folder was 26 GB. Is it on a per-workspace level?

**Reply by bobbrow** [reply]
@joebed, what you're reporting sounds like a bug. We don't keep a cache of that size per workspace if the cache path is the same for all workspaces. Do you see several files with a total size > 5GB, or is it one massive file?  If you delete the files in the ipch folder and resume working normally, are you able to get over the limit again?

**Reply by joebed** [reply]
I had 49 directories where the biggest one was 2 GB

**Reply by bobbrow** [reply]
If you still have the directory, can you share a directory listing (redact any personal info)?  `ls -l` or `dir /s`
I want to see if anything is abnormal with the file information.

**Reply by joebed** [reply]
sorry, already deleted it and its been working fine. I think it is related to opening a different directory in vscode:

To give some color on this, I'm using this with the Remote - SSH extension.

I had 1 directory in ~/.cache/vscode-cpptools/ipch, but when I opened a subdirectory of my project (via `code` command), I get another directory in the ipch folder. Did it again with a separate project directory with similar results. When I open a project directory I have previously opened, then no new directories are made in ipch.

**Reply by bobbrow** [reply]
Ok. Please let us know if you see this again.

**Reply by bobbrow** [reply]
Thank you for reporting this.  I discovered a regression with the implementation and we will get a fix out very soon.

**Reply by bobbrow** [reply]
Fixed in 1.19.5

**Reply by HIllya51** [reply]
> Yes. It's this one: ![image](https://private-user-images.githubusercontent.com/12818240/307403488-fd3b4e7a-799d-413c-bd9f-7b12dee2acba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg1NzMyMzMsIm5iZiI6MTc0ODU3MjkzMywicGF0aCI6Ii8xMjgxODI0MC8zMDc0MDM0ODgtZmQzYjRlN2EtNzk5ZC00MTNjLWJkOWYtN2IxMmRlZTJhY2JhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTMwVDAyNDIxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlYzU5NmQ1YWYzODIzZWQzNTg4NzMxZjUzYzg1NWY5YmIwZTdjZDg5MTU0MDdmNzUwYTA5YzAwMTZlZGMzODAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.V6sSXNyA1thUYyyHlIjU2GH1phOArvo0w_nRJusNnzY) 

Is this setting broken or is there a bug? I set it to 0 or a very small number, but it still keeps generating a lot of very large cache files💔

**Reply by sean-mcmanus** [reply]
@HIllya51 The setting only applies to ipch files and .db files -- what files are you seeing? We're not aware of a bug. Setting it to 0 should disable the feature.

**Comment by knalexschuetz** [comment]
Got the same problem:
vscode-cpptools/ipch is completely spammed with entries and eating a LOT of diskspace.
![image](https://github.com/microsoft/vscode-cpptools/assets/87977164/41872b6f-b1b2-41a4-943a-ed8875b065e8)

**Reply by bobbrow** [reply]
@knalexschuetz would it be possible to share a recursive listing?  Maybe move the `ipch` folder to `/tmp` first so that you don't have to manually redact all the personal information.  You should be able to drag/drop a txt file into the comment window to upload it.

**Reply by bobbrow** [reply]
Thank you for reporting this.  I discovered a regression with the implementation and we will get a fix out very soon.

**Reply by bobbrow** [reply]
Fixed in 1.19.5

**Comment by nelsonlachini** [comment]
Hi, I also found out my vscode-cpptools/ipch directory is filling a considerable amount of disk space, here is the recursive listing with sizes. The sizeable files correspond to .cpp files in my workspace. I'm using vscode-insiders 1.87.0 on a Mac BigSur 11.7.10.

[ipch_ls.txt](https://github.com/microsoft/vscode-cpptools/files/14480739/ipch_ls.txt)

**Reply by bobbrow** [reply]
Thank you for reporting this.  I discovered a regression with the implementation and we will get a fix out very soon.

**Reply by bobbrow** [reply]
Fixed in 1.19.5

**Reply by mimeau** [reply]
Thank you for the quick fix.

**Comment by miudr** [comment]
Hello,

In addition to the icph issue, I see in the vscode-cpptools/ folder in my home dir that a new object is created after each reopening of the project window. 
A very simple VSCode window close and reopen will trigger a new object to be created. In all cases, I open the same project and I use an SSH connection to the dev server.
![image](https://github.com/microsoft/vscode-cpptools/assets/84024197/e4afc831-c955-497c-9819-35f410cbc174)

It seems that it is related to the parsing of the workspace. What is strange is that the new object(created after the reopening goes up to 256 MB, then settles to 127MB as the others). 
![image](https://github.com/microsoft/vscode-cpptools/assets/84024197/367abb9c-e4e9-4e0a-a558-d434a99c5ba6)

Is this an expected behavior? If so, can it be disabled?

Thank you.

LE: forgot to mention, I am using v1.20.5.

**Comment by JeWaVe** [comment]
Well that folder took 249G on my system. Maybe implementing some kind of cache eviction policy would be a good thing

**Reply by tomachinz** [reply]
Yeah I reckon. Mines at 6.5 GB. No idea what to do about it.

**Comment by thomthom** [comment]
Can the location of this cache be configured? I'd prefer to have this of a dedicated cache HD rather my system HD.

(Mine as atm 12.2GB)

**Reply by sean-mcmanus** [reply]
@thomthom C_Cpp.intelliSenseCachePath (for ipch files).
